app/src/main/java/ch/hikemate/app/model/route/HikesViewModel.kt

```diff
@@ -0,0 +1,1166 @@
+  }
+
+  private suspend fun computeDetailsForAsync(
+      hikeId: String,
+      onSuccess: () -> Unit,
+      onFailure: () -> Unit
+  ) =
+      withContext(dispatcher) {
+        // Retrieve the current value of the hike and note what computations are required
+        val (performRequest, waypoints, elevation) =
+            checkDetailsCanBeRetrievedForAsync(hikeId, onSuccess, onFailure)
+
+        if (!performRequest) {
+          return@withContext
+        }
+
+        // Perform the actual computations
+        val distance: Double
+        val elevationGain: Double
+        val estimatedTime: Double
+        val difficulty: HikeDifficulty
+        try {
+          distance = RouteUtils.computeTotalDistance(waypoints)
+          elevationGain = RouteUtils.calculateElevationGain(elevation)
+          estimatedTime = RouteUtils.estimateTime(distance, elevationGain)
+          difficulty = RouteUtils.determineDifficulty(distance, elevationGain)
+        } catch (e: Exception) {
+          Log.e(LOG_TAG, "Error while computing hike details", e)
+
+          // Set the detail properties back to NotRequested
+          _hikesMutex.withLock {
+            val hikeFlow = _hikeFlowsMap[hikeId] ?: return@withLock
+            val updatedHike =
+                hikeFlow.value.copy(
+                    distance = DeferredData.NotRequested,
+                    elevationGain = DeferredData.NotRequested,
+                    estimatedTime = DeferredData.NotRequested,
+                    difficulty = DeferredData.NotRequested,
+                )
+            hikeFlow.value = updatedHike
+
+            // Update the selected hike if necessary
+            updateSelectedHike()
+          }
+
+          onFailure()
+          return@withContext
+        }
+
+        // Update the hike in the list
+        var success = false
+        _hikesMutex.withLock {
+          val hikeFlow = _hikeFlowsMap[hikeId] ?: return@withLock
+          val updatedHike =
+              hikeFlow.value.copy(
+                  distance = DeferredData.Obtained(distance),
+                  elevationGain = DeferredData.Obtained(elevationGain),
+                  estimatedTime = DeferredData.Obtained(estimatedTime),
+                  difficulty = DeferredData.Obtained(difficulty))
+          hikeFlow.value = updatedHike
+
+          // Update the selected hike if necessary
+          updateSelectedHike()
+
+          success = true
+        }
+
+        if (success) {
+          onSuccess()
+        } else {
+          onFailure()
+        }
+      }
+
+  /**
+   * Helper function for [computeDetailsForAsync].
+   *
+   * Checks that the required data ([Hike.waypoints] and [Hike.elevation]) are available, and that a
+   * request is not already ongoing. Returns the data if it is available, and indicates whether a
+   * request should be performed.
+   *
+   * If a request should be performed, this function will also mark the data as requested.
```

Comment:
Fair, the naming of the method might be a problem. It marks the data as requested because it does not make sense to check whether the data has been requested and then release the lock, making the check potentially invalid.

I renamed the function to `detailsPreRequestOperations`. I'm open to other ideas though.